libcst/codemod/visitors/_apply_type_annotations.py

```diff
@@ -0,0 +1,558 @@
+    import_names = set()
+    for _import in imports:
+        if isinstance(_import, cst.Import):
+            import_names.update(get_import_alias_names(_import.names))
+        else:
+            names = _import.names
+            if not isinstance(names, cst.ImportStar):
+                import_names.update(get_import_alias_names(names))
+    return import_names
+
+
+class FunctionAnnotation(NamedTuple):
+    parameters: cst.Parameters
+    returns: Optional[cst.Annotation]
+
+
+class ImportStatement(NamedTuple):
+    module: Union[cst.Name, cst.Attribute]
+    names: Set[str]
+
+
+class TypeCollector(cst.CSTVisitor):
+    """
+    Collect type annotations from a stub module.
+    """
+
+    def __init__(self, existing_imports: Set[str]) -> None:
+        # Qualifier for storing the canonical name of the current function.
+        self.qualifier: List[str] = []
+        # Store the annotations.
+        self.function_annotations: Dict[str, FunctionAnnotation] = {}
+        self.attribute_annotations: Dict[str, cst.Annotation] = {}
+        self.imports: Dict[str, ImportStatement] = {}
+        self.existing_imports: Set[str] = existing_imports
+        self.class_definitions: Dict[str, cst.ClassDef] = {}
+
+    def visit_ClassDef(self, node: cst.ClassDef) -> None:
+        self.qualifier.append(node.name.value)
+        self.class_definitions[node.name.value] = node
+
+    def leave_ClassDef(self, original_node: cst.ClassDef) -> None:
+        self.qualifier.pop()
+
+    def visit_FunctionDef(self, node: cst.FunctionDef) -> bool:
+        self.qualifier.append(node.name.value)
+        returns = node.returns
+        if returns is not None:
+            return_annotation = self._create_import_from_annotation(returns)
+            parameter_annotations = self._import_parameter_annotations(node.params)
+            self.function_annotations[".".join(self.qualifier)] = FunctionAnnotation(
+                parameters=parameter_annotations, returns=return_annotation
+            )
+        # pyi files don't support inner functions, return False to stop the traversal.
+        return False
+
+    def leave_FunctionDef(self, original_node: cst.FunctionDef) -> None:
+        self.qualifier.pop()
+
+    def visit_AnnAssign(self, node: cst.AnnAssign) -> bool:
+        name = _get_name_or_attribute_as_string(node.target)
+        if name is not None:
+            self.qualifier.append(name)
+        annotation_value = self._create_import_from_annotation(node.annotation)
+        self.attribute_annotations[".".join(self.qualifier)] = annotation_value
+        return True
+
+    def leave_AnnAssign(self, original_node: cst.AnnAssign) -> None:
+        self.qualifier.pop()
+
+    def visit_ImportFrom(self, node: cst.ImportFrom) -> None:
+        module = node.module
+        names = node.names
+        if module is None or isinstance(names, cst.ImportStar):
+            return
+        value = module.value
+        name = (
```

Comment:
I'm not sure the types check out here, I think just `name = _get_name_or_attribute_as_string(module)` might work.  See https://libcst.readthedocs.io/en/latest/nodes.html#libcst.ImportFrom